### PR TITLE
Rewrite v3

### LINKED ISSUE #19: Lighthouse report
Beano told me to put this here. I've compiled the results of lighthouse into a selection of bullet points.

# Performance
image

# Load times

I have put the issue and next to it the amount of time that could be saved (estimate) by making the change, changes are ranked by speed increase.

## Optimisations

* Use WebP or JPEG 2000 over PNG/JPEG for improved compression rates and faster load times (8.85s)
* Properly size images, store in the repo at the size they will be displayed to remove unnecessary loading (3.9s)
* Defer offscreen images - lazy load images not displayed immediately (1.76s)
* Minify JavaScript (1.15s)
* Remove unused CSS (1.15s)
* Eliminate render blocking resources - some resources block the first paint of the page, deliver critical JS/CSS inline and defer all non-critical (1.12s)

## Issues

* Use the CSS `font-display` feature to ensure text is visible during webfont loading.
* Serve static assets with an efficient cache policy, 40 resources are being used with inefficient caching
* Minimise work done on the main thread, make user of JavaScript workers

# Best Practices
* Make touch and wheel event listeners `passive` to improve the scroll performance on the page
* Links to cross-origin destinations are unsafe, add `rel="noopener"` or `rel="noreferrer"` to any external links to improve performance and prevent security vulnerabilities
* Image display dimensions should match the natural aspect ratio

Note: These are all recommendations, time increases are estimates by the software. Some of these issues could be invalid/unfixable.

### PR DESCRIPTION
**Pull request by Sh3llcod3**
Dear Team CD,

This is the long overdue rewrite version 3 of our main site. It has many added features, but most importantly, it has more regressions than I can count.

There comes a time in one's life when they have to face certain facts and come to terms with it. I have done the same.

I want to start off by saying that I am not a web-dev, neither do I have any tangible experience in web development. The main reason I created this site (aside from the obvious contribution to the community), is so that I could learn more about web development. Clearly, my approach of diving in head first with a material framework in order to achieve a good-looking site was beyond the scope and scale of my capabilities. I sincerely regret doing this and not learning the fundamentals of responsive design or JavaScript first. 

In this rewrite, I have done my best to address the main issue: slow load times. To tackle the problem I have identified the underlying issue, which is the costly system of creating the site's content using a set of JavaScript functions and HTML templates. This has mostly remedied the load time issue on top of which I have significantly reduced the number and size of image assets to further speed up load. This should close #19.

This is not perfect however and has taken our `index.html` page to exceed 1100+ lines (which is why I wrote the js system, since that could render the entire site in 50 or so lines once the functions were setup). Additionally, the quality of the visuals has taken a large hit and the site now largely feels empty.

I have also re-done the responsive grid system, which means it should scale to almost 100% of all available viewports out there in the world. The drawback, once again, is the fact that I have had to utilise a fixed grid system. No longer can it scale up to take advantage of the entire screen width, it is now capped at a fixed resolution based on a Google site. This may add to the site feeling empty, but that very much depends on the viewport's resolution. It does mean, that the site can now scale down properly to mobile and tablet viewports.

I have also added a dark theme option (in my opinion it looks better than the light theme and it should be made the default. I can do this if this gains enough support and make the light theme optional, a bit like how Discord handles theme). I've also added a `more` button to the projects tab which gets the `README.md` files in the root of the repositories and displays that after being converted by a markdown to html library. I strongly regret adding this option and it doesn't quite appear like I visualised it. Again, this is something I can remove if people want.

Anyhow, this site has cost me valuable time, which I should have spent revising. I feel that there was no reason to rewrite this in the first place. Nonetheless, there is no use contemplating over my poor decisions, I shall leave that to myself.

With that being said, there is two things I ask from the community. 

Firstly, I desperately seek honest feedback, since one cannot improve without making mistakes and learning from those mistakes. I can't improve if you tell me the site looks alright. I can only make it better if you point out my mistakes and show me what I could have done instead.

Secondly, the main reason I am putting out this PR is because for now I am throwing in the towel for a bit. This sort of monolithic chunky PRs rewriting the site isn't something that I can sustain for long. It has worn me out. Being (generally) the sole developer has meant I have had to make many decisions regarding design and layout which I didn't feel comfortable making. I have no knowledge of web-dev, neither do I have any experience (as I have stated before).

Therefore, it would only be wise to ask the community to help with the site. Many people have shown interest in helping out and improving the site, and I sincerely thank them in advance for any contributions they may be thinking of implementing.

I am in no hurry to merge. I can wait as long as it takes to achieve a reasonable look, functionality and feature parity with the old site before this is deployed to master. 

Simply put, I need help.

Yours sincerely, 
Sh3llcod3.

### CODE REVIEW

**Comment by NicholasG04** [comment]
Amazing rewrite! Definitely getting there. I think I can see some nice improvements and will take a look later.

Here's what I've just seen:

First page isn't the best use of width space I don't think
Navigation bar could be on every page maybe?
Disclaimer has no 'nice' route to get back.
Make external links open in a new tab please

**Comment by Sh3llcod3** [comment]
@NicholasG04 I agree with all these points, and I will be working on a fix in due course. Thanks :+1:

**Comment by NicholasG04** [comment]
Start a rewrite project, begin adding some points to that can be worked on and we can split them if you want :)

**Comment by Sh3llcod3** [comment]
> Start a rewrite project, begin adding some points to that can be worked on and we can split them if you want :)

Done!

**Comment by Bottersnike** [comment]
Visually, looks pretty good. I have just two little gripes though:

- A "clear cookies" checkbox in the settings a.) should really be a button not a checkbox, but more importantly b.) is pointless. The only cokies appear to be site theme and dismissing the cokkie message. If a user really for any reason wanted to clear those, they could, and should, do it manually, because there's really no reason to ever need to.
- When switching to tabs for the first time (esp the soundboard) it can take almost a second or two at times to load the new tab. I've not looked at the source yet to see why, but it might be worth seeing if it can be optimized out.

**Comment by Bottersnike** [comment]
I lied. Got another gripe. 
image
This text here is really thin, bordering on the there-isn't-actually-any-infill-just-antialiasing-artifacting level, which can make it quite hard to read at times. I think this might just be due to how Roboto works though, because it's rendering at a weight of 400 as it should. As such, this might be unfixable.

**Comment by Sh3llcod3** [comment]
@Bottersnike I have seen and acknowledged these issues and I am working on a patch.

**Comment by Sh3llcod3** [comment]
@Bottersnike @picapi You should now have access to the project board.

**Comment by Bottersnike** [comment]
Am I good to go through and shrink some of the large images?

**Comment by Sh3llcod3** [comment]
@Bottersnike Yes, please feel free to do so. However, I'm working on some patches, so your branch might go out of sync.

**Comment by Sh3llcod3** [comment]
@thebeanogamer @teamshortcut Since the camps are fast approaching, can you take a look at this please?

If there are any changes, I will make them as soon as you want.

File: index.html [hunk 1]
```diff
@@ -289,11 +131,11 @@ <h6 class="mdc-list-group__subheader">Friends of CD</h6>
   <header class="mdc-top-app-bar mdc-top-app-bar--fixed">
     <div class="mdc-top-app-bar__row">
       <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start">
-        <a class="material-icons mdc-top-app-bar__navigation-icon">menu</a>
-        <span class="mdc-top-app-bar__title">Community Projects</span>
+        <button class="material-icons mdc-top-app-bar__navigation-icon mdc-icon-button">menu</button>
+        <span class="mdc-top-app-bar__title">CD Community</span>
```
Comment (thebeanogamer) [inline index.html]:
```suggestion
        <span class="mdc-top-app-bar__title">Cyber Discovery Community</span>
```

File: index.html [hunk 2]
```diff
@@ -220,66 +62,66 @@ <h6 class="mdc-drawer__subtitle">Have you registered yet?</h6>
...
           <i class="material-icons mdc-list-item__graphic" aria-hidden="true">link</i>
           <span class="mdc-list-item__text">GitHub Page</span>
         </a>
-        <a class="mdc-list-item" href="https://play.google.com/store/apps/details?id=com.danielmilnes.cyberdiscovery">
+        <a class="mdc-list-item" href="https://play.google.com/store/apps/details?id=com.danielmilnes.cyberdiscovery" target="_blank">
           <i class="material-icons mdc-list-item__graphic" aria-hidden="true">get_app</i>
           <span class="mdc-list-item__text">Download App</span>
         </a>
-        <a class="mdc-list-item" href="https://cyber-discovery-app.firebaseapp.com/">
+        <a class="mdc-list-item" href="https://cyber-discovery-app.firebaseapp.com/" target="_blank">
```
Comment (thebeanogamer) [inline index.html]:
```suggestion
        <a class="mdc-list-item" href="https://app.cyberdiscoverycommunity.uk/" target="_blank">
```

File: index.html [hunk 3]
```diff
@@ -350,36 +197,980 @@ <h6 class="mdc-list-group__subheader">Friends of CD</h6>
...
+                          Fork Project
+                        </span>
+                      </li>
+                      <li class="mdc-list-item" role="menuitem" onclick="loadNewPage('https://github.com/CyberDiscovery/Cyber-Discovery-App-Website/issues');">
+                        <span class="mdc-list-item__text">
+                          <i class="material-icons project_card_menu_icons">view_list</i>
+                          View Issues
+                        </span>
+                      </li>
+                      <li class="mdc-list-item" role="menuitem" onclick="loadNewPage('https://cyber-discovery-app.firebaseapp.com/');">
```
Comment (thebeanogamer) [inline index.html]:
```suggestion
                      <li class="mdc-list-item" role="menuitem" onclick="loadNewPage('https://app.cyberdiscoverycommunity.uk/');">
```

Comment (joker314) [reply index.html]: Also, this appears to be semantically a hyperlink. Thus, it should probably have the appropriate ARIA role (and I think it isn't a menuitem, especially because this appears to be a navigation segment of the page - which from a brief search should be handled with its specific ARIA role).

The way I normally see navigation hyperlinks is with a DOM structure of NAV > LI > A. Of course, it doesn't matter how we implement the navigation section so long as it is accessible, and proper use of ARIA will help with that.

I haven't had a look at the rewrite but it looks really thorough and big and amazing, thank you so much!

File: Site_Pages/About_Page.html [hunk 4]
```diff
@@ -132,6 +139,30 @@
               </span>
               <span class="mdc-list-item__meta material-icons" aria-hidden="true" title="View GitHub Profile" onclick="loadNewPage('https://github.com/lightspeedana/')">star</span>
```
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply Site_Pages/About_Page.html]: Yep, done :+1:

File: index.html [hunk 5]
```diff
@@ -350,36 +197,980 @@ <h6 class="mdc-list-group__subheader">Friends of CD</h6>
...
+          </div>
+        </div>
+        <div class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-phone mdc-layout-grid__cell--span-4-tablet">
+          <div class="mdc-card mdc-elevation-transition projects-tab-card colored-card-bg">
+            <div class="padded-projects-card">
+              <h2 class="projects-card-title mdc-typography mdc-typography--headline3 light-dark-text">Welcome!</h2>
+            </div>
+            <div class="projects-card-secondary mdc-typography mdc-typography--body1 light-dark-text welcome-text-card">We're a passionate growing Discord community of students who are taking part in the UK's Cyber Discovery programme, part of HM Government's Cyber Schools initiative and the SANS CyberStart family of programmes. Find like-minded people, seek pointers and guidance on challenges, take part in Capture the flag events and more! If you're taking part in a related programme in the US, we welcome you as well! Make sure you join our server! While you're here, feel free to check out some of our cool open-source projects and consider lending us some of your amazing skills by contributing!</div>
```
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]: Absolutely, I'm not good at writing :eyes:

File: index.html [hunk 6]
```diff
@@ -350,36 +197,980 @@ <h6 class="mdc-list-group__subheader">Friends of CD</h6>
...
+              </div>
+            </div>
+          </div>
+        </div>
       </div>
     </div>
   </div>
   <div class="generic-tab-item layout-tab-item" id="Tab-3">
-    <div class="layout-grid-desktop mdc-layout-grid" id="App-Grid">
+    <div class="layout-grid-desktop mdc-layout-grid" id="Soundboard-Grid">
```
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]: Ok, I'll do that :airplane:

File: index.html [hunk 7]
```diff
@@ -350,36 +197,980 @@ <h6 class="mdc-list-group__subheader">Friends of CD</h6>
...
       <div class="mdc-layout-grid__inner" id="Grid-Tab-3">
       </div>
     </div>
   </div>
   <div class="generic-tab-item layout-tab-item" id="Tab-4">
-    <div class="layout-grid-desktop mdc-layout-grid" id="Websites-Grid">
+    <div class="layout-grid-desktop mdc-layout-grid" id="Events-Grid">
```
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]: @thebeanogamer You're right! I did notice that the formatting is not the same, I'll change it to have cards instead.

File: index.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]: Done!

File: index.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]: VC recordings? What VC recordings? I don't see them anywhere? Oh wait... :eyes:

File: index.html [hunk 8]
```diff
@@ -350,36 +197,980 @@ <h6 class="mdc-list-group__subheader">Friends of CD</h6>
...
+              <span class="mdc-typography--headline5">Lancaster, from the 30th July to the 1st August,</span>
+            </li>
+            <li class="centered-list-item">
+              <span class="mdc-typography--headline5">London, from the 6th August to the 8th August.</span>
+            </li>
+          </ul>
+          <h4 class="mdc-typography--headline4 medium-attention-text centered-item red-events-text">Do I need to journey there and back every day?</h4>
+          <h5 class="mdc-typography--headline5 medium-attention-text centered-item">Nope! We'll be provided with university accommodation for the two nights, with food and other necessities provided.</h5>
+          <h4 class="mdc-typography--headline4 medium-attention-text centered-item red-events-text">How do I know if I'm invited?</h4>
+          <h5 class="mdc-typography--headline5 medium-attention-text centered-item">Cyber Discovery has sent out emails to invite people to attend! If you've not recieved an email yet, don't give up hope! Emails regarding Elite will come from <b>elite@joincyberdiscovery.com</b> - make sure that the email doesn't end up in the spam folder!</h5>
```
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]: *rip*

File: index.html [hunk 9]
```diff
@@ -350,36 +197,980 @@ <h6 class="mdc-list-group__subheader">Friends of CD</h6>
...
+          <h5 class="mdc-typography--headline5 medium-attention-text centered-item">Yep! Alongside the SANS training at the event, you'll also have a chance to take a GIAC certification for that course after the event, which shows you know your stuff!</h5>
+          <h4 class="mdc-typography--headline4 medium-attention-text centered-item red-events-text">Do I need a parent?</h4>
+          <h5 class="mdc-typography--headline5 medium-attention-text centered-item">You do need a parent, or other authorized guardian, to sign you in and out at the start and end of the camp; but they don't need to stick around for the entire 6 days.</h5>
+          <h4 class="mdc-typography--headline4 medium-attention-text centered-item red-events-text">Any other questions?</h4>
+          <h5 class="mdc-typography--headline5 medium-attention-text centered-item">If the questions relate to personal circumstances, contact Cyber Discovery at <b>elite@joincyberdiscovery.com</b>, and they should be able to help! If it's more general, you can ask on our community discord server!</h5>
```
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]:
```suggestion
          <h5 class="mdc-typography--headline5 medium-attention-text centered-item">If the questions relate to personal circumstances, contact Cyber Discovery at <b>elite@joincyberdiscovery.com</b>, and they should be able to help! If it's more general, you can ask on our <a href="https://discord.gg/Kf8n5rT">community Discord server!</a></h5>
```

File: index.html [hunk 10]
```diff
@@ -350,36 +197,980 @@ <h6 class="mdc-list-group__subheader">Friends of CD</h6>
...
     <div class="layout-grid-desktop mdc-layout-grid" id="Resources-Grid">
       <div class="mdc-layout-grid__inner" id="Grid-Tab-5">
+        <div class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-phone mdc-layout-grid__cell--span-4-tablet" id="resources-intro-msg">
+          <div class="mdc-card mdc-elevation-transition resources-tab-card">
+            <div class="padded-resource-card">
+                <h2 class="resource-card-title mdc-typography mdc-typography--headline6">Hey There!</h2>
+                <h3 class="resource-card-subtitle mdc-typography mdc-typography--subtitle2">by Our Community</h3>
+            </div>
+            <div class="resource-card-secondary mdc-typography mdc-typography--body2">
+              Here are a group of resources you can use to enhance your cyber security knowledge and skill, regardless of your participation in Cyber Discovery.
```
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]: @thebeanogamer 

By two wide, do you mean that I should make it span the entire top of the section? Like this?

Demo

File: index.html (hunk 1 above)
Comment (Sh3llcod3) [reply index.html]: @thebeanogamer I am happy to make this change just like the other changes, but please note that doing this means that for people on mobile devices with < 5.5 inch screens, the text will be truncated and they will see some dots.

Demo2

File: index.html (hunk 3 above)
Comment (Sh3llcod3) [reply index.html]: @joker314 You know, you're right. I am confused by the documentation. Perhaps its best to remove the `role=` attribute here? I mean an earlier version of the framework used this and I haven't updated it since and they've removed all references to this attribute. :confused: 

https://material.io/develop/web/components/lists/

https://github.com/material-components/material-components-web/tree/master/packages/mdc-list

**Comment by Sh3llcod3** [comment]
@teamshortcut I've applied your changes and as you and @thebeanogamer wanted, the VC recordings are now non-existent.

@thebeanogamer Could you please take another look if possible? I've almost applied all your fixes :tada:

File: index.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (thebeanogamer) [reply index.html]: Yeah, I think that makes is more clearly seperate.

File: index.html (hunk 1 above)
Comment (thebeanogamer) [reply index.html]: That's... less than ideal

Comment (Sh3llcod3) [reply index.html]: Ok, I'll do it nonetheless! :+1:

File: index.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]: Ok, done! :+1:

File: index.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sh3llcod3) [reply index.html]: Its done, it now has cards. Hopefully it doesn't look too ugly now :thinking:

**Comment by Sh3llcod3** [comment]
~Massive IP dox there bois, what I am ever going to do now?~ :smile:

**Comment by Sh3llcod3** [comment]
Thanks guys! Let's merge this and see where this goes! :+1:
